### ISSUE TITLE
[Bug]: Inline styles blocked by CSP  `style-src` header

### ISSUE THREAD

**Issue by TannerS** [carbon-design-system/carbon#12580]
### Package

@carbon/react

### Browser

Chrome, Firefox

### Package version

^1.5.0

### React version

v17

## Background

IBM console / cloud has started to implement the CSP header ( https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP ). Part of that is it blocks inline styles, 

```
<style>
// css here
<style>

```
will be blocked unless it has a `nonce` value passed in by the consumer

where 

`<div style=....></div> `

will be blocked no matter what

this is problematic since all of ibm cloud ios moving to CSP, some inline styles (second example) will be blocked by CSP and break the component to some limit.

<details>
  <summary><h4>Here is some that I personally searched for and found in the carbon repo (click to expand)</h4></summary>
  
## DatePicker

with csp 
![image](https://user-images.githubusercontent.com/8866319/200667510-9f5aced0-bd96-4fd5-a319-44e79add850f.png)

without 
![image](https://user-images.githubusercontent.com/8866319/200667758-ca0e7795-3160-4a32-83f1-11b8d580f86f.png)

![image](https://user-images.githubusercontent.com/8866319/200658463-cc5cb843-bef7-4344-886b-dce02bd3cb41.png)

![image](https://user-images.githubusercontent.com/8866319/200658545-b825522d-3493-44bf-a2d6-a5de6f4011ec.png)

![image](https://user-images.githubusercontent.com/8866319/200658667-cc55a930-e770-421b-a365-d4b261208f3a.png)

![image](https://user-images.githubusercontent.com/8866319/200658727-65b580b7-fc03-4114-9ccb-1f78c55b89ca.png)

![image](https://user-images.githubusercontent.com/8866319/200658799-a45f81a1-8e19-4e5f-a970-ca2779f2e249.png)

![image](https://user-images.githubusercontent.com/8866319/200658851-dfa61843-b8c6-45ba-8325-9450ae49548d.png)

![image](https://user-images.githubusercontent.com/8866319/200658905-05e053c9-0720-416e-b15d-0d6def7b58e3.png)

This one for some reason doesnt give me a warning but if you look here theres a style property that gets added (i changed it to height to show an example)

![image](https://user-images.githubusercontent.com/8866319/200661034-fb91a035-6b67-488d-83a3-b4cbf2a947ef.png)

but csp can affect browsers differently so it should be something to look into 

## SkeletonText / Slider / ProgressBar 

These also have inline styles that may be blocked by csp so worth a look

## Other 

![image](https://user-images.githubusercontent.com/8866319/200667175-946f2033-764a-4d7f-bcf7-62f0717c4472.png)
</details>

### Suggested Severity

Severity 3 = User can complete task, and/or has a workaround within the user experience of a given component.

### Reproduction/example

N/A

### Steps to reproduce

I am testing on my local microservice for ibm cloud with csp to verify 

DatePicker

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

https://github.com/carbon-design-system/carbon/issues/12580#tasklist-block-a9785d44-ab13-438b-8bb3-d2ba230a2a72

**Comment by TannerS** [comment]
> Thanks for opening this with so much detail!
> 
> To summarize the situation (let me know if I miss anything):
> 
>     * When using a Content Security Policy (CSP) header with [`style-src` directive](https://content-security-policy.com/style-src/), inline styles are blocked
> 
>     * Allowing inline styles by [providing a `nonce`](https://content-security-policy.com/nonce/) is not an option because this method can only be applied to `<script>` or `<style>`, not `style=""`
> 
> 
> On our end, one potential way to fix this would be to reduce the usage of inline styles within the components. Unfortunately, the examples you list above are almost all coming from third party flatpickr library. The likelyhood that we're able to completely fix this is low until we revisit datepicker altogether, #11967.
> 
> In the meantime, it seems a decent workaround to this problem altogether is [using a hash](https://content-security-policy.com/examples/allow-inline-style/) to allow inline style attributes. It's safer than doing `unsafe-inline` and supported in all modern browsers. IIUC there's nothing that Carbon needs to do for this, you'd just need to add the CSP `unsafe-hashes` keyword to your `style-src` directive.

Sup man, so 

>When using a Content Security Policy (CSP) header with [style-src directive](https://content-security-policy.com/style-src/), inline styles are blocked

I think inline styles are blocked no matter what directives are defined, as long as csp is turned on 

>Allowing inline styles by [providing a nonce](https://content-security-policy.com/nonce/) is not an option because this method can only be applied to <script> or <style>, not style=""

correct, but nonce and hashing has some limits, (see below)

>In the meantime, it seems a decent workaround to this problem altogether is [using a hash](https://content-security-policy.com/examples/allow-inline-style/) to allow inline style attributes. It's safer than doing unsafe-inline and supported in all modern browsers. IIUC there's nothing that Carbon needs to do for this, you'd just need to add the CSP unsafe-hashes keyword to your style-src directive.

if i remember correctly, we cannot do any `unsafe-*` for cloud or else we will get pinged to fix it. To also add, if u were to go the hash route (which means any changes will break the hash) the consuming site also has to add it to there csp policy 

for example, most of console is using nonce, the consumer has to be able to generate it per request and attach it to all there stuff and add it to the policy

for hashes, if i remember correctly, you guys generate it but we have to add it for every single page that needs it csp policy which for most is a code change and any changes (or if they are generated differently) will break it 

i think that works for inline styles, i am a bit rusy on hashing but know some teams have tried it

**Comment by TannerS** [comment]
@tay1orjones  @tw15egan  besides the above, found a related issue https://github.com/carbon-design-system/carbon/issues/12386

**Comment by tay1orjones** [comment]
> we cannot do any `unsafe-*`

In that case, the workaround you'll need to explore is how to apply the required (inline) styles yourself via a class instead.

I think the primary things that can be tackled here are:

- Reduce the usage of inline styles within Carbon components
- Explore a fix or workaround for the inline styles applied by third party libraries
  - https://github.com/flatpickr/flatpickr/issues/2823

PRs to help out would be much appreciated

**Comment by TannerS** [comment]
Got ya, if I figure out more or a better way to help I will let y'all know!

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tay1orjones** [comment]
I've added a task list to the issue body containing the components using inline styling. It includes a task to determine if we can reduce/remove inline styling stemming from third party libraries like flatpickr (there may be others - downshift?)

**Comment by tay1orjones** [comment]
This would improve the situation greatly, but there's been no movement -https://github.com/w3c/webappsec-csp/issues/174

**Comment by TannerS** [comment]
The issue with this is all teams who use anything with unsafe have to get pce exception and get it fixed by a date

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tay1orjones** [comment]
Another way we could set styles dynamically would be to use CSSOM, which should be compatible with CSP.

Instead of rendering an inline style like this

```html
<div id="test" style="left: 300px;"/>
```

we could instead set it imperatively via

```js
document.getElementById('test').style.left = '300px';
```

**Comment by TannerS** [comment]
That may still trigger csp I think (think being the word) but diff will need to be tested

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by andy-blum** [comment]
Hey there! checking in from the carbon-for-ibmdotcom side of things and CSSOM _totally works_.

Per [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src#examples):

> Inline style attributes are also blocked:
>
> ```html
> <div style="display:none">Foo</div>
> ```
>
>As well as styles that are applied in JavaScript by setting the style attribute directly, or by setting [cssText](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/cssText):
>
>```js
> document.querySelector('div').setAttribute('style', 'display:none;');
> document.querySelector('div').style.cssText = 'display:none;';
>```
>
>However, styles properties that are set directly on the element's [style](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/style) property will not be blocked, allowing users to safely manipulate styles via JavaScript:
>
>```js
>document.querySelector('div').style.display = 'none';
>```

As an example, [here's a quick codepen](https://codepen.io/andy-blum/pen/ExRRRBj)

**Comment by dabrad26** [comment]
Hi. Was wondering if the groups on this ticket had any luck with SVGs and their inline stuff for their products. We are running into issues with the pictograms. So wondering if your products found a work around. Happy to connect via slack as well (@david)

**Comment by tay1orjones** [comment]
@andy-blum or @TannerS have either of you encountered and had to resolve issues with SVGs or the pictograms and CSP?

**Comment by andy-blum** [comment]
Which styles specifically are in an inline attribue? Could you move them to their own specific attributes?
 
 For example, instead of 
```
<path 
  id="path14"
  style="fill:#354a8e;fill-opacity:1;fill-rule:nonzero;stroke:none"
  d="L 4424.6,237.121"
>
```

You could do
```
<path
  id="path14"
  fill="#354a8e"
  fill-opacity="1"
  fill-rule="nonzero"
  stroke="none"
  d="L 4424.6,237.121"
>
```

[source](https://stackoverflow.com/a/71600358)

**Comment by dabrad26** [comment]
It seems to be however the SVG are made (some seem good. Some have issues). I assume most came from designers and just converted to react components.  

Updating all to be attributed instead of inline would be the solution. But a bit of a painful one for migrating all the SVGs.   So was just wondering if any better options appeared.

**Comment by andy-blum** [comment]
Not particularly. Inline styles are only allowable under a CSP if you know the exact markup beforehand and can pass a hash in with the CSP directive headers

**Comment by andy-blum** [comment]
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#hash-algorithm-base64-value

**Comment by JSoet** [comment]
Hi, I was just wondering what is the latest expected behaviour here after this issue was closed... I see there's some discussion earlier that maybe it wouldn't be possible to fix some of the sub-dependencies, but then I see in the associated PR it mentions "Remove inline styles from DatePicker arrow icons"... 

Is it expected now after this issue was closed that we shouldn't see any console warnings like `Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self' ` ? 

Or are there still inline styles used by the sub-components that would result in these warnings?

Edit: I originally asked the above because we were seeing warnings like that, but I hadn't realized we were on an older version of our software that didn't have the latest carbon updates. When we tested on the version with newer carbon, then we didn't get those warnings anymore.
